t2box_reader/t2box_parsers_unit.pas

```diff
@@ -0,0 +1,239 @@
+begin
+  BlockRead(box_file, bytes, SizeOf(bytes), number_read);
+  Value := ExtendedToDouble(bytes);
+  if logging then
+    T2box_log.Info('Parsed extended : ' + fldname + ' : ' + floattostr(Value));
+  Result := Value;
+end;
+
+function parse_float(var box_file: file; fldname: string): single;
+var
+  Value: single;
+  number_read: integer;
+begin
+  BlockRead(box_file, Value, SizeOf(Value), number_read);
+  if logging then
+    T2box_log.Info('Parsed float : ' + fldname + ' : ' + floattostr(Value));
+  Result := Value;
+end;
+
+function parse_integer(var box_file: file; fldname: string): integer;
+var
+  Value: int32;                      // integers are 32 bits in T2 box files
+  number_read: integer;
+begin
+  BlockRead(box_file, Value, SizeOf(Value), number_read);
+  if logging then
+    T2box_log.Info('Parsed integer : ' + fldname + ' : ' + inttostr(Value));
+  Result := Value;
+end;
+
+function parse_string(var box_file: file; fldname: string; strlen: integer): string;
+var
+  Value: string[255];
+  number_read: integer;
+begin
+  BlockRead(box_file, Value, strlen+1, number_read);
+  if logging then
+    T2box_log.Info('Parsed string : ' + fldname + ' : ' + Value);
+  Result := Value;
+end;
+
+// This function takes an array of bytes containing a bit pattern representing
+// an 'extended' float value.
+// The extended number is expected to be little-endian since it is assumed
+// to be produced by Templot2 on an Intel processor.
+function extendedToDouble(bytes: TextBytes): double;
+  //const
+var
+  rslt: double;
+  mantissa: Qword;
+  exponent: integer;
+  sign: byte;
+begin
+  mantissa := Pqword(@bytes[0])^;         //first get the mantissa
+
+  exponent := (Puint16(@bytes[8]))^;  // next get sign+exponent ...
+  sign := exponent shr 15;             // ... split off the sign ...
+  exponent := exponent and $7fff;      // ... and drop it from the exponent
+  case exponent of
+    0:
+      rslt := 0;
+    $7fff:        // TODO : This should interpret all sorts of odd values
+                  // (although not needed for Templot2 files)
+    begin
+      writeln('=== INVALID FLOAT - $7fff  ============>');
+      writeln('=== exponent = ', exponent);
+      //dump(Tbytes(bytes), 0, 10);
+      rslt := 0; // WRONG!!!  This should deal with various forms of infinity etc
+                 // (although not needed for Templot2 files)
+    end;
+    else begin
+      exponent := exponent - 16383;               // un-bias the exponent
+      if exponent < -1022 then begin
+        writeln('INVALID FLOAT - UNDERFLOW ============>');
+        writeln('=== exponent = ', exponent);
+        //dump(bytes, 0, 10);
+        rslt := 0;
+      end
+      else
+      if exponent > 1023 then begin
+        writeln('INVALID FLOAT - OVERFLOW =============>');
+        writeln('=== exponent = ', exponent);
+        //dump(bytes, 0, 10);
+        rslt := 0;
+      end
+      else
+        try
+          rslt := mantissa / intpower(2, 63 - exponent);
```

Comment:
Consider using TDoubleHelper.BuildUp() procedure.

Reply:
Thanks Alistair.
I will definitely use this in phase II. 
What reinvented wheels?

Reply:
The 'exponent' value here is a binary exponent, a bit pattern straight from the extended float. 

Unfortunately, the BuildUp procedure requires a decimal exponent.  Converting between the two would add extra, unnecessary code.

For efficiency we could bit-twiddle a double from the parts we have, but with a loss of clarity.  Unless we have a performance problem the code is probably better left as-is.

Reply:
Ok, fair enough to not use the BuildUp function, but 2 suggestions:
1) I looked at this for a while to work out what it was doing, I think it would be clearer if expressed as
```
rslt := mantissa * intpower(2, exponent - 63);
```
2) Create some unit tests...

Reply:
1. Yes, it is not obvious. The logic was that the mantissa in a normalised float (which is the only kind we need consider) is a 64 bit binary number with an implied binary point one digit in from the left. However, we pull it as a 64 bit integer - ie too big by a factor of 2^63.  Thus to get the 'real' mantissa, we need to shift it right 63 bits - same as dividing by 2^63.  However, we also have an exponent, which represents multiplying by 2^exponent - in other words the exponent needs to be extracted from the 63.  I did try rewriting it the way you suggest, but once I had this logic in my brain I couldn't see it as better any other way.  Anyway ... I have replaced this with 'bit twiddling, which came out much clearer than I expected, and feels much more "honest".

2. Absolutely! I have extracted the extendedToDouble function into a new unit "extended_utils" with tests.